// treedocumentmodel/src/TreedocumentmodelPackageLiterals.ts
export class TreedocumentmodelPackageLiterals {
  public static TREENODE: number = 0;
  public static TREENODE_FEATURE_COUNT: number = 7;
  public static TREENODE_OPERATION_COUNT: number = 0;

  public static TREE_NODE__ID: number = 0;
  public static TREE_NODE__TYPE: number = 1;
  public static TREE_NODE__NAME: number = 2;
  public static TREE_NODE__DOCUMENTATION: number = 3;
  public static TREE_NODE__CHILDREN: number = 4;
  public static TREE_NODE__PARENT: number = 5;
  public static TREE_NODE__ICON_OVERRIDE: number = 6;

  public static TREEDOCUMENT: number = 1;
  public static TREEDOCUMENT_FEATURE_COUNT: number = 4;
  public static TREEDOCUMENT_OPERATION_COUNT: number = 0;

  public static TREE_DOCUMENT__ROOTS: number = 0;
  public static TREE_DOCUMENT__ICONS: number = 1;
  public static TREE_DOCUMENT__ACTIONS: number = 2;
  public static TREE_DOCUMENT__HIERARCHY: number = 3;

  public static HIERARCHYCONFIGURATION: number = 2;
  public static HIERARCHYCONFIGURATION_FEATURE_COUNT: number = 2;
  public static HIERARCHYCONFIGURATION_OPERATION_COUNT: number = 0;

  public static HIERARCHY_CONFIGURATION__ALLOWED_CHILDREN: number = 0;
  public static HIERARCHY_CONFIGURATION__ROOTS: number = 1;

  public static NODETYPETONODETYPESMAP: number = 3;
  public static NODETYPETONODETYPESMAP_FEATURE_COUNT: number = 2;
  public static NODETYPETONODETYPESMAP_OPERATION_COUNT: number = 0;

  public static NODE_TYPE_TO_NODE_TYPES_MAP__KEY: number = 0;
  public static NODE_TYPE_TO_NODE_TYPES_MAP__VALUE: number = 1;

  public static ACTION: number = 4;
  public static ACTION_FEATURE_COUNT: number = 3;
  public static ACTION_OPERATION_COUNT: number = 0;

  public static ACTION__ID: number = 0;
  public static ACTION__NAME: number = 1;
  public static ACTION__ICON: number = 2;

  public static ACTIONCONFIGURATION: number = 5;
  public static ACTIONCONFIGURATION_FEATURE_COUNT: number = 3;
  public static ACTIONCONFIGURATION_OPERATION_COUNT: number = 0;

  public static ACTION_CONFIGURATION__AVAILABLE_ACTIONS: number = 0;
  public static ACTION_CONFIGURATION__DEFAULT_ACTIONBAR_ACTIONS: number = 1;
  public static ACTION_CONFIGURATION__NODE_ACTIONS: number = 2;

  public static ACTIONIDTONODETYPEMAP: number = 6;
  public static ACTIONIDTONODETYPEMAP_FEATURE_COUNT: number = 2;
  public static ACTIONIDTONODETYPEMAP_OPERATION_COUNT: number = 0;

  public static ACTION_ID_TO_NODE_TYPE_MAP__KEY: number = 0;
  public static ACTION_ID_TO_NODE_TYPE_MAP__VALUE: number = 1;

  public static NODETYPETOICONDATAURIMAP: number = 10;
  public static NODETYPETOICONDATAURIMAP_FEATURE_COUNT: number = 2;
  public static NODETYPETOICONDATAURIMAP_OPERATION_COUNT: number = 0;

  public static NODE_TYPE_TO_ICON_DATA_URI_MAP__KEY: number = 0;
  public static NODE_TYPE_TO_ICON_DATA_URI_MAP__VALUE: number = 1;

  public static NODEICON: number = 11;
  public static NODEICON_FEATURE_COUNT: number = 1;
  public static NODEICON_OPERATION_COUNT: number = 0;

  public static NODE_ICON__ICONS: number = 0;

  public static ACTIONEVENT: number = 12;
  public static ACTIONEVENT_FEATURE_COUNT: number = 2;
  public static ACTIONEVENT_OPERATION_COUNT: number = 0;

  public static ACTION_EVENT__TARGET_NODES: number = 0;
  public static ACTION_EVENT__ACTION: number = 1;

  public static ACTIONID: number = 7;

  public static NODETYPE: number = 8;

  public static ICONDATAURI: number = 9;
}
